src/mscorlib/src/System/ReadOnlySpan.cs

```diff
@@ -0,0 +1,269 @@
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <typeparamref name="T"/> is reference type or contains pointers and hence cannot be stored in unmanaged memory.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        [CLSCompliant(false)]
+        public unsafe ReadOnlySpan(void* ptr, int length)
+        {
+            if (JitHelpers.ContainsReferences<T>())
+                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
+            if (length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _rawPointer = (IntPtr)ptr;
+            _length = length;
+        }
+
+        /// <summary>
+        /// An internal helper for creating spans. Not for public use.
+        /// </summary>
+        internal ReadOnlySpan(ref T ptr, int length)
+        {
+            JitHelpers.SetByRef(out _rawPointer, ref ptr);
+            _length = length;
+        }
+
+        public static implicit operator ReadOnlySpan<T>(Span<T> slice)
+        {
+            return new ReadOnlySpan<T>(ref JitHelpers.GetByRef<T>(ref slice._rawPointer), slice.Length);
+        }
+
+        public static implicit operator ReadOnlySpan<T>(T[] array)
+        {
+            return new ReadOnlySpan<T>(array);
+        }
+
+        public static implicit operator ReadOnlySpan<T>(ArraySegment<T> arraySegment)
+        {
+            return new ReadOnlySpan<T>(arraySegment.Array, arraySegment.Offset, arraySegment.Count);
+        }
+
+        public int Length
+        {
+            get { return _length; }
+        }
+
+        public static ReadOnlySpan<T> Empty
+        {
+            get { return default(ReadOnlySpan<T>); }
+        }
+
+        public bool IsEmpty
+        {
+            get { return _length == 0; }
+        }
+
+        /// <summary>
+        /// Fetches the element at the specified index.
+        /// </summary>
+        /// <exception cref="System.IndexOutOfRangeException">
+        /// Thrown when the specified <paramref name="index"/> is not in range (&lt;0 or &gt;&eq;Length).
+        /// </exception>
+        public T this[int index]
+        {
+            get
+            {
+                if ((uint)index >= (uint)_length)
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+
+                return JitHelpers.AddByRef(ref JitHelpers.GetByRef<T>(ref _rawPointer), index);
+            }
+        }
+
+        /// <summary>
+        /// Copies the contents of this span into a new array.  This heap
+        /// allocates, so should generally be avoided, however is sometimes
+        /// necessary to bridge the gap with APIs written in terms of arrays.
+        /// </summary>
+        public T[] ToArray()
+        {
+            var destination = new T[_length];
+            TryCopyTo(destination);
```

Comment:
What if it fails? A `Try` pattern implies it can fail and a success value is returned, which is not used here. I'd expect it to be either used or have an inline comment if it's sure that it **can't** fail.